jubilant/_helpers.py

```diff
@@ -0,0 +1,18 @@
+from ._types import Status
+
+
+def any_error(status: Status) -> bool:
+    """Return true if any application or unit in *status* is in "error" status."""
```

Comment:
It probably can't be helped in this library, since it's really a Juju issue, but I feel like "status" (the Juju status of a whole model that you get with `juju status`) and "status" (the string that's the "workload-status") and "status" (the string that's the agent (?) status) and "status" (the string that's the application status) and "status" (the string that's the machine status) and "status" (the string that's the model status) is going to be rather confusing.

Do all the string ones at least have the same possible set of values? That would be better, although I still think that the "juju status" and "status as a string" are going to be confusing to refer to.

Reply:
Yeah, this is confusing. I think application and unit status have the same (limited) set of values. The agent status (from the "juju-status" field in the JSON) has a bunch of other values like "idle". But I think the context is going to help solve this: here, the `any_error` and similar functions will only be looking at application and unit statuses. And the `is_app_active` functions are clearly application (we might have `is_unit_active` and co as well). Then if we need `is_agent_idle` variants they'll be clear too.

Reply:
Two thoughts: If we're using strings for statuses, it would probably be worth using a `Literal` for the typing. Have we considered using an `Enum` instead though? Or something like `Literal['active', 'etc'] | UnitStatus` ... even with a `StrEnum` (or `UnitStatus(str, Enum)`) where `'active' == UnitStatus.ACTIVE`, you can get value in avoiding errors since `UnitStatus.ACTIVE is not AppStatus.ACTIVE`

I agree with @tonyandrewmeyer about the `juju status` being confused with specific status values. Maybe we'd want something like `JujuState` instead of `Status`, with the argument name then being `state`?

Reply:
Note that I'm not exposing strings as statuses in this public API yet, the statuses are in the method names, like `is_app_active`. I definitely want to avoid "stringly typed" errors in the main API. There is a status string in the big `Status` object (the "current" field). I actually think enums are a bit of a pain, so I'd lean towards keeping that a string but typing it with `Literal['active', ...]` as you suggest.

Reply:
I agree that it's nice if we can avoid exposing string statuses directly, though I guess the `current` field pokes a hole in that already? I do wonder if it would be worth doing enums right if we have to expose a status ... but maybe there's a way to avoid exposing current at all? If you had `is_active` etc on the `ApplicationStatus` then that could replace `current`

Reply:
I'm going to keep `current` and the like as it mirrors the incoming object from Juju. However, per below, I'm adding `ApplicationStatus.is_<status>` helper properties and removed the `is_app_active()` methods on the Status object. I think that's better -- thanks!